StripeFinancialConnections/StripeFinancialConnections/Source/Native/Consent/ConsentDataSource.swift

```
@@ -25,24 +41,80 @@ final class ConsentDataSourceImplementation: ConsentDataSource {
         clientSecret: String,
-        analyticsClient: FinancialConnectionsAnalyticsClient
+        analyticsClient: FinancialConnectionsAnalyticsClient,
+        elementsSessionContext: ElementsSessionContext?
     ) {
         self.manifest = manifest
         self.consent = consent
         self.merchantLogo = merchantLogo
         self.apiClient = apiClient
         self.clientSecret = clientSecret
         self.analyticsClient = analyticsClient
+        self.elementsSessionContext = elementsSessionContext
     }
 
-    func markConsentAcquired() -> Promise<FinancialConnectionsSessionManifest> {
-        return apiClient.markConsentAcquired(clientSecret: clientSecret)
+    func markConsentAcquired() -> Future<ConsentAcquiredResult> {
+        return apiClient.markConsentAcquired(clientSecret: clientSecret).chained { [weak self] manifest in
+            guard let self, manifest.shouldLookupConsumerSession, let email else {
+                let result = ConsentAcquiredResult(manifest: manifest)
+                return Promise(value: result)
+            }
+
+            let promise = Promise<ConsentAcquiredResult>()
+
+            apiClient.consumerSessionLookup(
+                emailAddress: email,
+                clientSecret: clientSecret,
+                sessionId: manifest.id,
+                emailSource: .customerObject,
+                useMobileEndpoints: manifest.verified,
+                pane: .consent
+            ).observe { lookupResult in
+                switch lookupResult {
+                case .success(let response):
+                    let result = ConsentAcquiredResult(
+                        manifest: manifest,
+                        consumerSession: response.consumerSession,
+                        publishableKey: response.publishableKey
+                    )
+                    promise.resolve(with: result)
+                case .failure:
+                    let result = ConsentAcquiredResult(manifest: manifest)
```

Just to confirm, we're ok failing silently here?

---

Yes. It just means that looking up the session failed, so we’d navigate to the manifest’s `nextPane`.

If that’s `linkLogin` (like it is for IBP today), the user would be recognized after their email is prefilled (basically today‘s experience).